### ISSUE TITLE
Unhashable type: 'ValidateCallWrapper' when creating a model which has a @validate_call wrapped field annotation.

### ISSUE THREAD

**Issue by spinks** [pydantic/pydantic#8094]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

A new issue as of v2.5.0. 

When creating a model:
- which contains a field whose type annotation is a union of a `typing.Annotated` and any other type, 
- if a `pydantic.functional_validators` (`BeforeValidator`, `AfterValidator`, etc) is used within that `Annotated`
- and the function passed to that validator has been wrapped with a `validate_call` decorator

then a `TypeError: unhashable type: 'ValidateCallWrapper'` is thrown.

If a field looks like:
- `name: Annotated[str, AfterValidator(validation_function)] | None` it will error if that validation function is `@validate_call` wrapped
  - `None` can be any type.
- An unwrapped validation function will not error in this case.
- A type annotation which is not a Union of an `Annotated` with a `functional_validator` + `validate_call` decorated  and another type will not error.

The error trace  for example code is as follows:
```
  File "/Users/benspinks/simple.py", line 11, in <module>
    class User(BaseModel):
  File "/Users/benspinks/simple.py", line 13, in User
    name: Annotated[str, AfterValidator(lower_case_func)] | None
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
  File "/Users/benspinks/miniforge3/lib/python3.11/typing.py", line 1375, in __or__
    return Union[self, right]
           ~~~~~^^^^^^^^^^^^^
  File "/Users/benspinks/miniforge3/lib/python3.11/typing.py", line 358, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/benspinks/miniforge3/lib/python3.11/typing.py", line 481, in __getitem__
    return self._getitem(self, parameters)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/benspinks/miniforge3/lib/python3.11/typing.py", line 695, in Union
    parameters = _remove_dups_flatten(parameters)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/benspinks/miniforge3/lib/python3.11/typing.py", line 326, in _remove_dups_flatten
    return tuple(_deduplicate(params))
                 ^^^^^^^^^^^^^^^^^^^^
  File "/Users/benspinks/miniforge3/lib/python3.11/typing.py", line 301, in _deduplicate
    all_params = set(params)
                 ^^^^^^^^^^^
  File "/Users/benspinks/miniforge3/lib/python3.11/typing.py", line 2151, in __hash__
    return hash((self.__origin__, self.__metadata__))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 3, in __hash__
TypeError: unhashable type: 'ValidateCallWrapper'
``` 

### Example Code

```Python
from typing import Annotated

from pydantic import AfterValidator, BaseModel, validate_call


@validate_call
def lower_case_func(s: str) -> str:
    return s.lower()


class User(BaseModel):
    name: Annotated[str, AfterValidator(lower_case_func)] | None
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.5.0
        pydantic-core version: 2.14.1
          pydantic-core build: profile=release pgo=true
                 install path: /Users/benspinks/venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.6 (main, Oct  2 2023, 13:45:54) [Clang 15.0.0 (clang-1500.0.40.1)]
                     platform: macOS-14.1-arm64-arm-64bit
             related packages: typing_extensions-4.8.0
```

**Comment by sydney-runkle** [comment]
@spinks,

Thanks a bunch for reporting this 👍. We'll look into a fix to push out with a `v2.5.1` release soon. I can confirm this looks like a bug 🐛.

**Comment by sydney-runkle** [comment]
Addressed here: 5321fd5d
